app/com/arpnetworking/metrics/portal/alerts/impl/CacheActor.java

```diff
@@ -0,0 +1,184 @@
+import akka.actor.ActorRef;
+import akka.actor.Props;
+import akka.actor.Status;
+import akka.pattern.Patterns;
+import com.arpnetworking.metrics.incubator.PeriodicMetrics;
+import com.arpnetworking.steno.Logger;
+import com.arpnetworking.steno.LoggerFactory;
+import com.google.common.collect.Maps;
+
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * A actor that acts as a simple key-value store. You can use this as a cluster-wide cache
+ * by starting as a cluster singleton, or as a building-block for a more complex cache
+ * by using it as a member within a ClusterSharding pool.
+ *
+ * @param <K> The type of key.
+ * @param <V> The type of the associated value.
+ *
+ * @author Christian Briones (cbriones at dropbox dot com)
+ */
+public final class CacheActor<K, V> extends AbstractActor {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CacheActor.class);
+
+    private final PeriodicMetrics _metrics;
+    private final String _cacheName;
+    private final HashMap<K, V> _cache;
+
+    private CacheActor(final String cacheName, final PeriodicMetrics metrics) {
+        _cacheName = cacheName;
+        _metrics = metrics;
+        _cache = Maps.newHashMap();
+    }
+
+    /**
+     * Props to construct a new instance of {@code CacheActor}.
+     *
+     * @param cacheName The name of the cache.
+     * @param metrics A metrics instance to record against.
+     * @return props to instantiate the actor
+     */
+    public static Props props(final String cacheName, final PeriodicMetrics metrics) {
+        return Props.create(CacheActor.class, () -> new CacheActor<>(cacheName, metrics));
+    }
+
+    /**
+     * Retrieve a the value associated with this key.
+     *
+     * @param ref The CacheActor ref.
+     * @param key The key to retrieve.
+     * @param timeout The operation timeout.
+     * @param <K> The type of key.
+     * @param <V> The type of value.
+     * @return A CompletionStage which contains the value, if any.
+     */
+    @SuppressWarnings("unchecked")
+    public static <K, V> CompletionStage<Optional<V>> get(final ActorRef ref, final K key, final Duration timeout) {
+        return Patterns.askWithReplyTo(
+            ref,
+            replyTo -> new CacheGet<>(key, replyTo),
+            timeout
+        ).thenApply(resp -> (Optional<V>) resp);
+    }
+
+    /**
+     * Put a Key-Value pair into the cache.
+     *
+     * @param ref The CacheActor ref.
+     * @param key The key to update.
+     * @param value The associated value.
+     * @param timeout The operation timeout.
+     * @param <K> The type of key.
+     * @param <V> The type of value.
+     * @return A completion stage to await for the write to complete.
+     */
+    public static <K, V> CompletionStage<Void> put(final ActorRef ref, final K key, final V value, final Duration timeout) {
+        return Patterns.askWithReplyTo(
```

Comment:
It would not be, this API is just for interacting with a single actor. that is, I wouldn't expect to be able to pass the proxy actor ref here if we were doing more complex things like replicating the cache.

I don't need any of that here though so maybe I can just rename the class and/or remove that suggestion in the docstring?

Reply:
Yeah I'm fine with that.

A generic cache abstracting its deployment would have been really sweet. However, it's out of scope.